{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3d5c14af-d537-4d47-90b9-234e1451f412",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import gzip\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0166038a-ecc9-4246-9b7c-768c3c9413f2",
   "metadata": {},
   "source": [
    "# 1. Loading genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6e7d8d93-e683-4179-ae54-0024d8abf76b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to the genome FASTA file\n",
    "GENOME_PATH = \"hg38.fa\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "652e5a45-3d7b-42de-98a9-50ba5987ce94",
   "metadata": {},
   "outputs": [],
   "source": [
    "class DataSource:\n",
    "    \"\"\"\n",
    "    Sourced from https://github.com/meuleman/SynthSeqs/blob/main/make_data/source.py\n",
    "    A class to represent a source of reference genome data.\n",
    "\n",
    "    Attributes\n",
    "    ----------\n",
    "    raw_data : str\n",
    "        The raw genomic data in FASTA format.\n",
    "    filepath : str\n",
    "        The path to the file containing the genomic data.\n",
    "    \n",
    "    Methods\n",
    "    -------\n",
    "    data():\n",
    "        Returns the raw genomic data.\n",
    "    \"\"\"\n",
    "    def __init__(self, data, filepath):\n",
    "        self.raw_data = data\n",
    "        self.filepath = filepath\n",
    "\n",
    "    @property\n",
    "    def data(self):\n",
    "        return self.raw_data\n",
    "\n",
    "class DataSource:\n",
    "    # Sourced from https://github.com/meuleman/SynthSeqs/blob/main/make_data/source.py\n",
    "\n",
    "    def __init__(self, data, filepath):\n",
    "        self.raw_data = data\n",
    "        self.filepath = filepath\n",
    "\n",
    "    @property\n",
    "    def data(self):\n",
    "        return self.raw_data\n",
    "\n",
    "\n",
    "class ReferenceGenome(DataSource):\n",
    "    \"\"\"\n",
    "    A class for quickly loading and querying the reference genome.\n",
    "    \n",
    "    Methods\n",
    "    -------\n",
    "    from_path(path):\n",
    "        Class method to create a ReferenceGenome instance from a FASTA file.\n",
    "    \n",
    "    from_dict(data_dict):\n",
    "        Class method to create a ReferenceGenome instance from a dictionary.\n",
    "    \n",
    "    genome():\n",
    "        Property that returns the genome data.\n",
    "    \n",
    "    sequence(chrom, start, end):\n",
    "        Returns the sequence of a specific chromosomal region.\n",
    "    \"\"\"\n",
    "\n",
    "    @classmethod\n",
    "    def from_path(cls, path):\n",
    "        \"\"\"\n",
    "        Creates a ReferenceGenome instance from a FASTA file.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        path : str\n",
    "            The path to the FASTA file.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        ReferenceGenome\n",
    "            An instance of ReferenceGenome with the genome data loaded from the file.\n",
    "        \"\"\"\n",
    "        genome_dict = {record.id: str(record.seq).upper() for record in SeqIO.parse(path, \"fasta\")}\n",
    "        return cls(genome_dict, path)\n",
    "\n",
    "    @classmethod\n",
    "    def from_dict(cls, data_dict):\n",
    "        \"\"\"\n",
    "        Creates a ReferenceGenome instance from a dictionary.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        data_dict : dict\n",
    "            A dictionary where keys are chromosome names and values are sequences.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        ReferenceGenome\n",
    "            An instance of ReferenceGenome with the genome data loaded from the dictionary.\n",
    "        \"\"\"\n",
    "        return cls(data_dict, filepath=None)\n",
    "\n",
    "    @property\n",
    "    def genome(self):\n",
    "        \"\"\"\n",
    "        Returns the genome data.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        dict\n",
    "            A dictionary where keys are chromosome names and values are sequences.\n",
    "        \"\"\"\n",
    "        return self.data\n",
    "\n",
    "    def sequence(self, chrom, start, end):\n",
    "        \"\"\"\n",
    "        Returns the sequence of a specific chromosomal region.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        chrom : str\n",
    "            The name of the chromosome.\n",
    "        start : int\n",
    "            The starting position of the sequence.\n",
    "        end : int\n",
    "            The ending position of the sequence.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        str\n",
    "            The DNA sequence from the specified chromosomal region.\n",
    "\n",
    "        Raises\n",
    "        ------\n",
    "        AssertionError\n",
    "            If the end position is out of the chromosome sequence range.\n",
    "        \"\"\"\n",
    "        chrom_sequence = self.genome[chrom]\n",
    "\n",
    "        assert end < len(chrom_sequence), (\n",
    "            f\"Sequence position bound out of range for chromosome {chrom}. \"\n",
    "            f\"{chrom} length {len(chrom_sequence)}, requested position {end}.\"\n",
    "        )\n",
    "        return chrom_sequence[start:end]\n",
    "\n",
    "\n",
    "genome = ReferenceGenome.from_path(GENOME_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cb031aeb-a9b7-4a93-b2a2-cf0fdf3fcf4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library order</th>\n",
       "      <th>Biosample name</th>\n",
       "      <th>Vocabulary representative</th>\n",
       "      <th>DCC Experiment ID</th>\n",
       "      <th>DCC Library ID</th>\n",
       "      <th>DCC Biosample ID</th>\n",
       "      <th>DCC File ID</th>\n",
       "      <th>Altius Aggregation ID</th>\n",
       "      <th>Altius Library ID</th>\n",
       "      <th>Altius Biosample ID</th>\n",
       "      <th>...</th>\n",
       "      <th>Library cleanup</th>\n",
       "      <th>DNaseI units/mL</th>\n",
       "      <th>Amount Nucleic Acid (ng)</th>\n",
       "      <th>Nuclei count</th>\n",
       "      <th>Protease inhibitor</th>\n",
       "      <th>Library sequencing date</th>\n",
       "      <th>Reads used</th>\n",
       "      <th>DCC SPOT score</th>\n",
       "      <th>Per-biosample peaks</th>\n",
       "      <th>DHSs in Index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GM06990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMQ</td>\n",
       "      <td>ENCLB435ZZZ</td>\n",
       "      <td>ENCBS057ENC</td>\n",
       "      <td>ENCFF983CTQ</td>\n",
       "      <td>AG5636</td>\n",
       "      <td>LN1203</td>\n",
       "      <td>DS7748</td>\n",
       "      <td>...</td>\n",
       "      <td>Sucrose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2009-02-23</td>\n",
       "      <td>142681590.0</td>\n",
       "      <td>0.6790</td>\n",
       "      <td>83639.0</td>\n",
       "      <td>82918.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>HepG2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000ENP</td>\n",
       "      <td>ENCLB480ZZZ</td>\n",
       "      <td>ENCBS114ENC</td>\n",
       "      <td>ENCFF419JVG</td>\n",
       "      <td>AG5635</td>\n",
       "      <td>LN1207</td>\n",
       "      <td>DS7764</td>\n",
       "      <td>...</td>\n",
       "      <td>Sucrose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2009-02-23</td>\n",
       "      <td>138826342.0</td>\n",
       "      <td>0.5858</td>\n",
       "      <td>89748.0</td>\n",
       "      <td>89235.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>hTH1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EQC</td>\n",
       "      <td>ENCLB591ZZZ</td>\n",
       "      <td>ENCBS345AAA</td>\n",
       "      <td>ENCFF575KOF</td>\n",
       "      <td>AG5634</td>\n",
       "      <td>LN1222</td>\n",
       "      <td>DS7840</td>\n",
       "      <td>...</td>\n",
       "      <td>Sucrose</td>\n",
       "      <td>6.0</td>\n",
       "      <td>534.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007-06-06</td>\n",
       "      <td>149158633.0</td>\n",
       "      <td>0.6470</td>\n",
       "      <td>94360.0</td>\n",
       "      <td>93665.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>Hela</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000ENO</td>\n",
       "      <td>ENCLB479ZZZ</td>\n",
       "      <td>ENCBS890POO</td>\n",
       "      <td>ENCFF503PAE</td>\n",
       "      <td>AG4219</td>\n",
       "      <td>LN1264</td>\n",
       "      <td>DS8200</td>\n",
       "      <td>...</td>\n",
       "      <td>new Sucrose</td>\n",
       "      <td>4.0</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007-08-24</td>\n",
       "      <td>23372724.0</td>\n",
       "      <td>0.6444</td>\n",
       "      <td>59098.0</td>\n",
       "      <td>59024.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>CACO2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMI</td>\n",
       "      <td>ENCLB422ZZZ</td>\n",
       "      <td>ENCBS391ENC</td>\n",
       "      <td>ENCFF977BRD</td>\n",
       "      <td>AG4218</td>\n",
       "      <td>LN1269</td>\n",
       "      <td>DS8235</td>\n",
       "      <td>...</td>\n",
       "      <td>Sucrose</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007-09-05</td>\n",
       "      <td>22760059.0</td>\n",
       "      <td>0.7190</td>\n",
       "      <td>29894.0</td>\n",
       "      <td>29724.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   library order Biosample name Vocabulary representative DCC Experiment ID  \\\n",
       "0            1.0        GM06990                       NaN       ENCSR000EMQ   \n",
       "1            2.0          HepG2                       NaN       ENCSR000ENP   \n",
       "2            3.0           hTH1                       NaN       ENCSR000EQC   \n",
       "3            4.0           Hela                       NaN       ENCSR000ENO   \n",
       "4            5.0          CACO2                       NaN       ENCSR000EMI   \n",
       "\n",
       "  DCC Library ID DCC Biosample ID  DCC File ID Altius Aggregation ID  \\\n",
       "0    ENCLB435ZZZ      ENCBS057ENC  ENCFF983CTQ                AG5636   \n",
       "1    ENCLB480ZZZ      ENCBS114ENC  ENCFF419JVG                AG5635   \n",
       "2    ENCLB591ZZZ      ENCBS345AAA  ENCFF575KOF                AG5634   \n",
       "3    ENCLB479ZZZ      ENCBS890POO  ENCFF503PAE                AG4219   \n",
       "4    ENCLB422ZZZ      ENCBS391ENC  ENCFF977BRD                AG4218   \n",
       "\n",
       "  Altius Library ID Altius Biosample ID  ... Library cleanup DNaseI units/mL  \\\n",
       "0            LN1203              DS7748  ...         Sucrose             NaN   \n",
       "1            LN1207              DS7764  ...         Sucrose             NaN   \n",
       "2            LN1222              DS7840  ...         Sucrose             6.0   \n",
       "3            LN1264              DS8200  ...     new Sucrose             4.0   \n",
       "4            LN1269              DS8235  ...         Sucrose             8.0   \n",
       "\n",
       "  Amount Nucleic Acid (ng) Nuclei count Protease inhibitor  \\\n",
       "0                       50          NaN                NaN   \n",
       "1                       50          NaN                NaN   \n",
       "2                    534.9          NaN                NaN   \n",
       "3                       50          NaN                NaN   \n",
       "4                        1          NaN                NaN   \n",
       "\n",
       "  Library sequencing date   Reads used DCC SPOT score Per-biosample peaks  \\\n",
       "0              2009-02-23  142681590.0         0.6790             83639.0   \n",
       "1              2009-02-23  138826342.0         0.5858             89748.0   \n",
       "2              2007-06-06  149158633.0         0.6470             94360.0   \n",
       "3              2007-08-24   23372724.0         0.6444             59098.0   \n",
       "4              2007-09-05   22760059.0         0.7190             29894.0   \n",
       "\n",
       "  DHSs in Index  \n",
       "0       82918.0  \n",
       "1       89235.0  \n",
       "2       93665.0  \n",
       "3       59024.0  \n",
       "4       29724.0  \n",
       "\n",
       "[5 rows x 38 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "The dataframe contains metadata related to biosamples and experimental details.\n",
    "It includes information such as library order, biosample names, various IDs from DCC and Altius,\n",
    "experimental protocols, biosample characteristics, and additional quantitative and qualitative metrics.\n",
    "\"\"\"\n",
    "DHS_Index_and_Vocabulary_metadata = pd.read_table('./DHS_Index_and_Vocabulary_metadata.tsv').iloc[:-1]\n",
    "DHS_Index_and_Vocabulary_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f9207a7a-483b-43b2-a15d-218e48d19ba5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['library order', 'Biosample name', 'Vocabulary representative',\n",
       "       'DCC Experiment ID', 'DCC Library ID', 'DCC Biosample ID',\n",
       "       'DCC File ID', 'Altius Aggregation ID', 'Altius Library ID',\n",
       "       'Altius Biosample ID', 'Replicate indicators', 'System', 'Subsystem',\n",
       "       'Organ', 'Biosample type', 'Biological state', 'Germ layer',\n",
       "       'Description ', 'Growth stage', 'Age', 'Sex', 'Ethnicity', 'Donor ID',\n",
       "       'Unique cellular condition', 'Used in Figure 1b', 'Biosample protocol',\n",
       "       'Experiment protocol', 'Library kit method', 'Library cleanup',\n",
       "       'DNaseI units/mL', 'Amount Nucleic Acid (ng)', 'Nuclei count',\n",
       "       'Protease inhibitor', 'Library sequencing date', 'Reads used',\n",
       "       'DCC SPOT score', 'Per-biosample peaks', 'DHSs in Index'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DHS_Index_and_Vocabulary_metadata.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e11ae75f-4a54-4418-8640-bc124ad43b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the NMF basis array from a .npy file\n",
    "basis_array = np.load('2018-06-08NC16_NNDSVD_Basis.npy')\n",
    "\n",
    "# Save the basis array to a CSV file for easier manipulation with pandas\n",
    "np.savetxt(\"2018-06-08NC16_NNDSVD_Basis.csv\", basis_array, delimiter=\",\")\n",
    "\n",
    "# Load the NMF basis array from the CSV file into a pandas DataFrame\n",
    "nmf_loadings = pd.read_csv('2018-06-08NC16_NNDSVD_Basis.csv', header=None)\n",
    "\n",
    "# Rename columns for clarity, creating columns C1 to C16 for the 16 components\n",
    "nmf_loadings.columns = ['C' + str(i) for i in range(1, 17)]\n",
    "\n",
    "\n",
    "# Join the metadata DataFrame with the NMF loadings DataFrame\n",
    "# This combines the metadata with the component presence matrix\n",
    "DHS_Index_and_Vocabulary_metadata = pd.concat([DHS_Index_and_Vocabulary_metadata, nmf_loadings], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e3df59f2-3d3c-400a-919c-7eb96ddcdc10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library order</th>\n",
       "      <th>Biosample name</th>\n",
       "      <th>Vocabulary representative</th>\n",
       "      <th>DCC Experiment ID</th>\n",
       "      <th>DCC Library ID</th>\n",
       "      <th>DCC Biosample ID</th>\n",
       "      <th>DCC File ID</th>\n",
       "      <th>Altius Aggregation ID</th>\n",
       "      <th>Altius Library ID</th>\n",
       "      <th>Altius Biosample ID</th>\n",
       "      <th>...</th>\n",
       "      <th>C7</th>\n",
       "      <th>C8</th>\n",
       "      <th>C9</th>\n",
       "      <th>C10</th>\n",
       "      <th>C11</th>\n",
       "      <th>C12</th>\n",
       "      <th>C13</th>\n",
       "      <th>C14</th>\n",
       "      <th>C15</th>\n",
       "      <th>C16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GM06990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMQ</td>\n",
       "      <td>ENCLB435ZZZ</td>\n",
       "      <td>ENCBS057ENC</td>\n",
       "      <td>ENCFF983CTQ</td>\n",
       "      <td>AG5636</td>\n",
       "      <td>LN1203</td>\n",
       "      <td>DS7748</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.102685</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.026774</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>HepG2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000ENP</td>\n",
       "      <td>ENCLB480ZZZ</td>\n",
       "      <td>ENCBS114ENC</td>\n",
       "      <td>ENCFF419JVG</td>\n",
       "      <td>AG5635</td>\n",
       "      <td>LN1207</td>\n",
       "      <td>DS7764</td>\n",
       "      <td>...</td>\n",
       "      <td>0.193557</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.074557</td>\n",
       "      <td>0.095928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.190564</td>\n",
       "      <td>0.416094</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>hTH1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EQC</td>\n",
       "      <td>ENCLB591ZZZ</td>\n",
       "      <td>ENCBS345AAA</td>\n",
       "      <td>ENCFF575KOF</td>\n",
       "      <td>AG5634</td>\n",
       "      <td>LN1222</td>\n",
       "      <td>DS7840</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>Hela</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000ENO</td>\n",
       "      <td>ENCLB479ZZZ</td>\n",
       "      <td>ENCBS890POO</td>\n",
       "      <td>ENCFF503PAE</td>\n",
       "      <td>AG4219</td>\n",
       "      <td>LN1264</td>\n",
       "      <td>DS8200</td>\n",
       "      <td>...</td>\n",
       "      <td>0.155545</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.407768</td>\n",
       "      <td>0.113676</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.420549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>CACO2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMI</td>\n",
       "      <td>ENCLB422ZZZ</td>\n",
       "      <td>ENCBS391ENC</td>\n",
       "      <td>ENCFF977BRD</td>\n",
       "      <td>AG4218</td>\n",
       "      <td>LN1269</td>\n",
       "      <td>DS8235</td>\n",
       "      <td>...</td>\n",
       "      <td>0.131876</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.936955</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   library order Biosample name Vocabulary representative DCC Experiment ID  \\\n",
       "0            1.0        GM06990                       NaN       ENCSR000EMQ   \n",
       "1            2.0          HepG2                       NaN       ENCSR000ENP   \n",
       "2            3.0           hTH1                       NaN       ENCSR000EQC   \n",
       "3            4.0           Hela                       NaN       ENCSR000ENO   \n",
       "4            5.0          CACO2                       NaN       ENCSR000EMI   \n",
       "\n",
       "  DCC Library ID DCC Biosample ID  DCC File ID Altius Aggregation ID  \\\n",
       "0    ENCLB435ZZZ      ENCBS057ENC  ENCFF983CTQ                AG5636   \n",
       "1    ENCLB480ZZZ      ENCBS114ENC  ENCFF419JVG                AG5635   \n",
       "2    ENCLB591ZZZ      ENCBS345AAA  ENCFF575KOF                AG5634   \n",
       "3    ENCLB479ZZZ      ENCBS890POO  ENCFF503PAE                AG4219   \n",
       "4    ENCLB422ZZZ      ENCBS391ENC  ENCFF977BRD                AG4218   \n",
       "\n",
       "  Altius Library ID Altius Biosample ID  ...        C7   C8        C9  \\\n",
       "0            LN1203              DS7748  ...  0.000000  0.0  0.000000   \n",
       "1            LN1207              DS7764  ...  0.193557  0.0  0.074557   \n",
       "2            LN1222              DS7840  ...  0.000000  0.0  0.000000   \n",
       "3            LN1264              DS8200  ...  0.155545  0.0  0.000000   \n",
       "4            LN1269              DS8235  ...  0.131876  0.0  0.000000   \n",
       "\n",
       "        C10  C11  C12       C13       C14  C15       C16  \n",
       "0  0.102685  0.0  0.0  0.026774  0.000000  0.0  0.000000  \n",
       "1  0.095928  0.0  0.0  3.190564  0.416094  0.0  0.000000  \n",
       "2  0.000000  0.0  0.0  0.000000  0.000000  0.0  0.000000  \n",
       "3  0.000000  0.0  0.0  0.407768  0.113676  0.0  2.420549  \n",
       "4  0.000000  0.0  0.0  0.936955  0.000000  0.0  0.000000  \n",
       "\n",
       "[5 rows x 54 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DHS_Index_and_Vocabulary_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "22029f92-f43d-47a7-8005-8c39bd107670",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the list of component column names\n",
    "COMPONENT_COLUMNS = [\n",
    "    'C1',\n",
    "    'C2',\n",
    "    'C3',\n",
    "    'C4',\n",
    "    'C5',\n",
    "    'C6',\n",
    "    'C7',\n",
    "    'C8',\n",
    "    'C9',\n",
    "    'C10',\n",
    "    'C11',\n",
    "    'C12',\n",
    "    'C13',\n",
    "    'C14',\n",
    "    'C15',\n",
    "    'C16',\n",
    "]\n",
    "\n",
    "# Create a new column 'component' in the metadata DataFrame\n",
    "# This column represents the index of the component with the highest value for each sample\n",
    "DHS_Index_and_Vocabulary_metadata['Component'] = (\n",
    "    DHS_Index_and_Vocabulary_metadata[COMPONENT_COLUMNS].idxmax(axis=1).apply(lambda x: int(x[1:]))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "55a12e98-091a-4702-a567-c593b637c35f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library order</th>\n",
       "      <th>Biosample name</th>\n",
       "      <th>Vocabulary representative</th>\n",
       "      <th>DCC Experiment ID</th>\n",
       "      <th>DCC Library ID</th>\n",
       "      <th>DCC Biosample ID</th>\n",
       "      <th>DCC File ID</th>\n",
       "      <th>Altius Aggregation ID</th>\n",
       "      <th>Altius Library ID</th>\n",
       "      <th>Altius Biosample ID</th>\n",
       "      <th>...</th>\n",
       "      <th>C8</th>\n",
       "      <th>C9</th>\n",
       "      <th>C10</th>\n",
       "      <th>C11</th>\n",
       "      <th>C12</th>\n",
       "      <th>C13</th>\n",
       "      <th>C14</th>\n",
       "      <th>C15</th>\n",
       "      <th>C16</th>\n",
       "      <th>Component</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GM06990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMQ</td>\n",
       "      <td>ENCLB435ZZZ</td>\n",
       "      <td>ENCBS057ENC</td>\n",
       "      <td>ENCFF983CTQ</td>\n",
       "      <td>AG5636</td>\n",
       "      <td>LN1203</td>\n",
       "      <td>DS7748</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.102685</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.026774</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>HepG2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000ENP</td>\n",
       "      <td>ENCLB480ZZZ</td>\n",
       "      <td>ENCBS114ENC</td>\n",
       "      <td>ENCFF419JVG</td>\n",
       "      <td>AG5635</td>\n",
       "      <td>LN1207</td>\n",
       "      <td>DS7764</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.074557</td>\n",
       "      <td>0.095928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.190564</td>\n",
       "      <td>0.416094</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>hTH1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EQC</td>\n",
       "      <td>ENCLB591ZZZ</td>\n",
       "      <td>ENCBS345AAA</td>\n",
       "      <td>ENCFF575KOF</td>\n",
       "      <td>AG5634</td>\n",
       "      <td>LN1222</td>\n",
       "      <td>DS7840</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>Hela</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000ENO</td>\n",
       "      <td>ENCLB479ZZZ</td>\n",
       "      <td>ENCBS890POO</td>\n",
       "      <td>ENCFF503PAE</td>\n",
       "      <td>AG4219</td>\n",
       "      <td>LN1264</td>\n",
       "      <td>DS8200</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.407768</td>\n",
       "      <td>0.113676</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.420549</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>CACO2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMI</td>\n",
       "      <td>ENCLB422ZZZ</td>\n",
       "      <td>ENCBS391ENC</td>\n",
       "      <td>ENCFF977BRD</td>\n",
       "      <td>AG4218</td>\n",
       "      <td>LN1269</td>\n",
       "      <td>DS8235</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.936955</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   library order Biosample name Vocabulary representative DCC Experiment ID  \\\n",
       "0            1.0        GM06990                       NaN       ENCSR000EMQ   \n",
       "1            2.0          HepG2                       NaN       ENCSR000ENP   \n",
       "2            3.0           hTH1                       NaN       ENCSR000EQC   \n",
       "3            4.0           Hela                       NaN       ENCSR000ENO   \n",
       "4            5.0          CACO2                       NaN       ENCSR000EMI   \n",
       "\n",
       "  DCC Library ID DCC Biosample ID  DCC File ID Altius Aggregation ID  \\\n",
       "0    ENCLB435ZZZ      ENCBS057ENC  ENCFF983CTQ                AG5636   \n",
       "1    ENCLB480ZZZ      ENCBS114ENC  ENCFF419JVG                AG5635   \n",
       "2    ENCLB591ZZZ      ENCBS345AAA  ENCFF575KOF                AG5634   \n",
       "3    ENCLB479ZZZ      ENCBS890POO  ENCFF503PAE                AG4219   \n",
       "4    ENCLB422ZZZ      ENCBS391ENC  ENCFF977BRD                AG4218   \n",
       "\n",
       "  Altius Library ID Altius Biosample ID  ...   C8        C9       C10  C11  \\\n",
       "0            LN1203              DS7748  ...  0.0  0.000000  0.102685  0.0   \n",
       "1            LN1207              DS7764  ...  0.0  0.074557  0.095928  0.0   \n",
       "2            LN1222              DS7840  ...  0.0  0.000000  0.000000  0.0   \n",
       "3            LN1264              DS8200  ...  0.0  0.000000  0.000000  0.0   \n",
       "4            LN1269              DS8235  ...  0.0  0.000000  0.000000  0.0   \n",
       "\n",
       "   C12       C13       C14  C15       C16 Component  \n",
       "0  0.0  0.026774  0.000000  0.0  0.000000         5  \n",
       "1  0.0  3.190564  0.416094  0.0  0.000000        13  \n",
       "2  0.0  0.000000  0.000000  0.0  0.000000         5  \n",
       "3  0.0  0.407768  0.113676  0.0  2.420549        16  \n",
       "4  0.0  0.936955  0.000000  0.0  0.000000         1  \n",
       "\n",
       "[5 rows x 55 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DHS_Index_and_Vocabulary_metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9727e27a-9457-4da2-971c-4433630eddfc",
   "metadata": {},
   "source": [
    "# Create sequence metadata frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1ff81e37-28b0-45cb-8d93-13a817ee1d80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This matrix represents the contribution of each of the 16 NMF components to each of the 3.6 million DHSs\n",
    "nmf_loadings = pd.read_csv('2018-06-08NC16_NNDSVD_Mixture.csv', header=None, names=COMPONENT_COLUMNS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c4877084-c538-43bd-8b2e-96e6e2db791d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3714439/3348375859.py:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  sequence_metadata = pd.read_table('./DHS_Index_and_Vocabulary_hg38_WM20190703.txt', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "# Load the metadata for the 3.6 million DHSs\n",
    "sequence_metadata = pd.read_table('./DHS_Index_and_Vocabulary_hg38_WM20190703.txt', sep='\\t')\n",
    "\n",
    "# Drop the 'component' column from the metadata DataFrame if it exists\n",
    "sequence_metadata = sequence_metadata.drop(columns=['component'], axis=1)\n",
    "\n",
    "# Merge the sequence metadata with the NMF loadings DataFrame\n",
    "# This combines the metadata for each DHS with its corresponding NMF component contributions\n",
    "df = pd.concat([sequence_metadata, nmf_loadings], axis=1, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "91befe6a-a336-4f4f-8884-47b4e2f1c3ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>identifier</th>\n",
       "      <th>mean_signal</th>\n",
       "      <th>numsamples</th>\n",
       "      <th>summit</th>\n",
       "      <th>core_start</th>\n",
       "      <th>core_end</th>\n",
       "      <th>C1</th>\n",
       "      <th>...</th>\n",
       "      <th>C7</th>\n",
       "      <th>C8</th>\n",
       "      <th>C9</th>\n",
       "      <th>C10</th>\n",
       "      <th>C11</th>\n",
       "      <th>C12</th>\n",
       "      <th>C13</th>\n",
       "      <th>C14</th>\n",
       "      <th>C15</th>\n",
       "      <th>C16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16140</td>\n",
       "      <td>16200</td>\n",
       "      <td>1.10011</td>\n",
       "      <td>0.129388</td>\n",
       "      <td>1</td>\n",
       "      <td>16170</td>\n",
       "      <td>16170.0</td>\n",
       "      <td>16170.0</td>\n",
       "      <td>0.001439</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>51868</td>\n",
       "      <td>52040</td>\n",
       "      <td>1.10021</td>\n",
       "      <td>0.080034</td>\n",
       "      <td>1</td>\n",
       "      <td>51970</td>\n",
       "      <td>51970.0</td>\n",
       "      <td>51970.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011431</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57280</td>\n",
       "      <td>57354</td>\n",
       "      <td>1.10025</td>\n",
       "      <td>0.273251</td>\n",
       "      <td>4</td>\n",
       "      <td>57350</td>\n",
       "      <td>57350.0</td>\n",
       "      <td>57350.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025745</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>66370</td>\n",
       "      <td>66482</td>\n",
       "      <td>1.10027</td>\n",
       "      <td>0.183716</td>\n",
       "      <td>8</td>\n",
       "      <td>66430</td>\n",
       "      <td>66410.0</td>\n",
       "      <td>66430.0</td>\n",
       "      <td>0.001336</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002904</td>\n",
       "      <td>0.001445</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003885</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>79100</td>\n",
       "      <td>79231</td>\n",
       "      <td>1.1003</td>\n",
       "      <td>0.113049</td>\n",
       "      <td>2</td>\n",
       "      <td>79150</td>\n",
       "      <td>79150.0</td>\n",
       "      <td>79150.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006965</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.001768</td>\n",
       "      <td>0.003912</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  start    end identifier  mean_signal  numsamples  summit  \\\n",
       "0    chr1  16140  16200    1.10011     0.129388           1   16170   \n",
       "1    chr1  51868  52040    1.10021     0.080034           1   51970   \n",
       "2    chr1  57280  57354    1.10025     0.273251           4   57350   \n",
       "3    chr1  66370  66482    1.10027     0.183716           8   66430   \n",
       "4    chr1  79100  79231     1.1003     0.113049           2   79150   \n",
       "\n",
       "   core_start  core_end        C1  ...        C7        C8        C9  \\\n",
       "0     16170.0   16170.0  0.001439  ...  0.000040  0.000000  0.000000   \n",
       "1     51970.0   51970.0  0.000000  ...  0.011431  0.000000  0.000000   \n",
       "2     57350.0   57350.0  0.000000  ...  0.000000  0.025745  0.000000   \n",
       "3     66410.0   66430.0  0.001336  ...  0.002904  0.001445  0.000000   \n",
       "4     79150.0   79150.0  0.000000  ...  0.006965  0.000000  0.000208   \n",
       "\n",
       "        C10       C11  C12       C13  C14  C15  C16  \n",
       "0  0.000000  0.000000  0.0  0.000000  0.0  0.0  0.0  \n",
       "1  0.000000  0.000000  0.0  0.000000  0.0  0.0  0.0  \n",
       "2  0.000000  0.000000  0.0  0.000000  0.0  0.0  0.0  \n",
       "3  0.000000  0.000000  0.0  0.003885  0.0  0.0  0.0  \n",
       "4  0.001768  0.003912  0.0  0.000000  0.0  0.0  0.0  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ce46132c-d100-4d89-817e-bb5ae3674777",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sequence_bounds(summit: int, start: int, end: int, length: int):\n",
    "    \"\"\"\n",
    "    Calculate the sequence coordinates (bounds) for a given DHS.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    summit : int\n",
    "        The summit position of the DHS.\n",
    "    start : int\n",
    "        The starting position of the DHS.\n",
    "    end : int\n",
    "        The ending position of the DHS.\n",
    "    length : int\n",
    "        The desired length of the sequence.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    tuple\n",
    "        A tuple containing the left and right bounds of the sequence.\n",
    "    \n",
    "    https://github.com/meuleman/SynthSeqs/blob/main/make_data/process.py\n",
    "    \"\"\"\n",
    "    half = length // 2\n",
    "\n",
    "    if (summit - start) < half:\n",
    "        return start, start + length\n",
    "    elif (end - summit) < half:\n",
    "        return end - length, end\n",
    "\n",
    "    return summit - half, summit + half\n",
    "\n",
    "def add_sequence_column(df: pd.DataFrame, genome, length: int):\n",
    "    \"\"\"\n",
    "    Query the reference genome for each DHS and add the raw sequences\n",
    "    to the dataframe.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        The dataframe of DHS annotations and NMF loadings.\n",
    "    genome : ReferenceGenome(DataSource)\n",
    "        A reference genome object to query for sequences.\n",
    "    length : int\n",
    "        The desired length of the DHS sequence.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        The dataframe with an added column 'sequence' containing the sequences for each DHS.\n",
    "    \n",
    "    https://github.com/meuleman/SynthSeqs/blob/main/make_data/process.py\n",
    "    \"\"\"\n",
    "    seqs = []\n",
    "    for rowi, row in df.iterrows():\n",
    "        # Determine the left and right bounds of the sequence\n",
    "        l, r = sequence_bounds(row['summit'], row['start'], row['end'], length)\n",
    "        \n",
    "        # Query the genome to get the sequence for the given chromosomal region\n",
    "        seq = genome.sequence(row['seqname'], l, r)\n",
    "        \n",
    "        # Append the sequence to the list\n",
    "        seqs.append(seq)\n",
    "\n",
    "    # Add the sequences as a new column in the dataframe\n",
    "    df['sequence'] = seqs\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2148b607-e56b-4381-8566-f821cbf137f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Recreating some of the columns from our original dataset\n",
    "\n",
    "# Determine the component with the highest value for each DHS and create a 'component' column\n",
    "# The component is determined by finding the column with the maximum value in the COMPONENT_COLUMNS\n",
    "df['component'] = df[COMPONENT_COLUMNS].idxmax(axis=1).apply(lambda x: int(x[1:]))\n",
    "\n",
    "# Calculate the proportion of the maximum component value to the sum of all component values for each DHS\n",
    "# This represents how dominant the highest component is relative to the others\n",
    "df['proportion'] = df[COMPONENT_COLUMNS].max(axis=1) / df[COMPONENT_COLUMNS].sum(axis=1)\n",
    "\n",
    "# Calculate the total signal for each DHS by multiplying the mean signal by the number of samples\n",
    "df['total_signal'] = df['mean_signal'] * df['numsamples']\n",
    "\n",
    "# Calculate the proportion again (this line seems redundant as it is the same as the previous calculation)\n",
    "df['proportion'] = df[COMPONENT_COLUMNS].max(axis=1) / df[COMPONENT_COLUMNS].sum(axis=1)\n",
    "\n",
    "# Create a unique identifier for each DHS by concatenating the chromosome name, start, end, and summit positions\n",
    "# The identifier is in the format 'seqname_start_end_summit'\n",
    "df['dhs_id'] = df[['seqname', 'start', 'end', 'summit']].apply(lambda x: '_'.join(map(str, x)), axis=1)\n",
    "\n",
    "# Calculate the width of each DHS by subtracting the start position from the end position\n",
    "df['DHS_width'] = df['end'] - df['start']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ad891dd9-9225-4a88-bbed-cf5e8e4f1816",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating sequence column\n",
    "df = add_sequence_column(df, genome, 200)\n",
    "\n",
    "# Changing seqname column to chr\n",
    "df = df.rename(columns={'seqname': 'chr'})\n",
    "\n",
    "# Reordering and unselecting columns\n",
    "df = df[\n",
    "    [\n",
    "        'dhs_id',\n",
    "        'chr',\n",
    "        'start',\n",
    "        'end',\n",
    "        'DHS_width',\n",
    "        'summit',\n",
    "        'numsamples',\n",
    "        'total_signal',\n",
    "        'component',\n",
    "        'proportion',\n",
    "        'sequence',\n",
    "    ]\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "81b80aec-ca21-4ea9-9925-54e7a2e48d44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dhs_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>DHS_width</th>\n",
       "      <th>summit</th>\n",
       "      <th>numsamples</th>\n",
       "      <th>total_signal</th>\n",
       "      <th>component</th>\n",
       "      <th>proportion</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_16140_16200_16170</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16140</td>\n",
       "      <td>16200</td>\n",
       "      <td>60</td>\n",
       "      <td>16170</td>\n",
       "      <td>1</td>\n",
       "      <td>0.129388</td>\n",
       "      <td>1</td>\n",
       "      <td>0.855153</td>\n",
       "      <td>CGGGCATCCTGTGTGCAGATACTCCCTGCTTCCTCTCTAGCCCCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_51868_52040_51970</td>\n",
       "      <td>chr1</td>\n",
       "      <td>51868</td>\n",
       "      <td>52040</td>\n",
       "      <td>172</td>\n",
       "      <td>51970</td>\n",
       "      <td>1</td>\n",
       "      <td>0.080034</td>\n",
       "      <td>7</td>\n",
       "      <td>0.973545</td>\n",
       "      <td>GGCGACCCAGCGAGACTCCGCCTCAAAAAAAAAAAAAGAAGATTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_57280_57354_57350</td>\n",
       "      <td>chr1</td>\n",
       "      <td>57280</td>\n",
       "      <td>57354</td>\n",
       "      <td>74</td>\n",
       "      <td>57350</td>\n",
       "      <td>4</td>\n",
       "      <td>1.093002</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>CTCAGTCATTCCGAACAATTCACACACTAAGATTACCCATGCTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_66370_66482_66430</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66370</td>\n",
       "      <td>66482</td>\n",
       "      <td>112</td>\n",
       "      <td>66430</td>\n",
       "      <td>8</td>\n",
       "      <td>1.469725</td>\n",
       "      <td>3</td>\n",
       "      <td>0.332213</td>\n",
       "      <td>ATATATAAATTATATAATATAATATATATTATATAATATAATATAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_79100_79231_79150</td>\n",
       "      <td>chr1</td>\n",
       "      <td>79100</td>\n",
       "      <td>79231</td>\n",
       "      <td>131</td>\n",
       "      <td>79150</td>\n",
       "      <td>2</td>\n",
       "      <td>0.226098</td>\n",
       "      <td>7</td>\n",
       "      <td>0.501840</td>\n",
       "      <td>CATTTCTCCAAGGAGGAAATACCAGAGTCAATTCACAACCACTGCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   dhs_id   chr  start    end  DHS_width  summit  numsamples  \\\n",
       "0  chr1_16140_16200_16170  chr1  16140  16200         60   16170           1   \n",
       "1  chr1_51868_52040_51970  chr1  51868  52040        172   51970           1   \n",
       "2  chr1_57280_57354_57350  chr1  57280  57354         74   57350           4   \n",
       "3  chr1_66370_66482_66430  chr1  66370  66482        112   66430           8   \n",
       "4  chr1_79100_79231_79150  chr1  79100  79231        131   79150           2   \n",
       "\n",
       "   total_signal  component  proportion  \\\n",
       "0      0.129388          1    0.855153   \n",
       "1      0.080034          7    0.973545   \n",
       "2      1.093002          8    1.000000   \n",
       "3      1.469725          3    0.332213   \n",
       "4      0.226098          7    0.501840   \n",
       "\n",
       "                                            sequence  \n",
       "0  CGGGCATCCTGTGTGCAGATACTCCCTGCTTCCTCTCTAGCCCCCA...  \n",
       "1  GGCGACCCAGCGAGACTCCGCCTCAAAAAAAAAAAAAGAAGATTGA...  \n",
       "2  CTCAGTCATTCCGAACAATTCACACACTAAGATTACCCATGCTAAA...  \n",
       "3  ATATATAAATTATATAATATAATATATATTATATAATATAATATAT...  \n",
       "4  CATTTCTCCAAGGAGGAAATACCAGAGTCAATTCACAACCACTGCA...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a72803f7-420b-4da8-82b7-21d6968dab54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the binary matrix\n",
    "# This matrix represents the presence (1) or absence (0) of DHS peaks in 733 biosamples\n",
    "binary_matrix = pd.read_table('./dat_bin_FDR01_hg38.txt', header=None)\n",
    "\n",
    "# Collecting names of cells into a list with format celltype_encodeID\n",
    "\n",
    "# Create a list of cell type identifiers by concatenating 'Biosample name' and 'DCC Library ID'\n",
    "# The resulting format for each cell type identifier is 'BiosampleName_DCCLibraryID'\n",
    "celltype_encodeID = [\n",
    "    row['Biosample name'] + \"_\" + row['DCC Library ID'] for _, row in DHS_Index_and_Vocabulary_metadata.iterrows()\n",
    "]\n",
    "\n",
    "# Renaming columns using celltype_encodeID list\n",
    "\n",
    "# Rename the columns of the binary matrix using the celltype_encodeID list\n",
    "# Each column now corresponds to a specific cell type and library ID combination\n",
    "binary_matrix.columns = celltype_encodeID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e7ed1cab-1714-4ec8-9a76-b0ea6e2a224f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GM06990_ENCLB435ZZZ</th>\n",
       "      <th>HepG2_ENCLB480ZZZ</th>\n",
       "      <th>hTH1_ENCLB591ZZZ</th>\n",
       "      <th>Hela_ENCLB479ZZZ</th>\n",
       "      <th>CACO2_ENCLB422ZZZ</th>\n",
       "      <th>CACO2_ENCLB423ZZZ</th>\n",
       "      <th>SKNSH_ENCLB585ZZZ</th>\n",
       "      <th>SKNSH_ENCLB586ZZZ</th>\n",
       "      <th>HMEC_ENCLB493ZZZ</th>\n",
       "      <th>GM12878_ENCLB441ZZZ</th>\n",
       "      <th>...</th>\n",
       "      <th>fKidney_ENCLB005SRL</th>\n",
       "      <th>fKidney_ENCLB704GMQ</th>\n",
       "      <th>fKidney_ENCLB759USM</th>\n",
       "      <th>fLung_ENCLB594BSZ</th>\n",
       "      <th>fKidney_ENCLB049MNH</th>\n",
       "      <th>fUmbilical_cord_ENCLB771UER</th>\n",
       "      <th>fBone_femur_ENCLB236BWV</th>\n",
       "      <th>fLiver_ENCLB638FEH</th>\n",
       "      <th>fPlacenta_ENCLB423VBC</th>\n",
       "      <th>fPlacenta_ENCLB711ZZZ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 733 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   GM06990_ENCLB435ZZZ  HepG2_ENCLB480ZZZ  hTH1_ENCLB591ZZZ  Hela_ENCLB479ZZZ  \\\n",
       "0                    0                  0                 0                 0   \n",
       "1                    0                  0                 0                 0   \n",
       "2                    0                  0                 0                 0   \n",
       "3                    0                  0                 0                 0   \n",
       "4                    0                  0                 0                 0   \n",
       "\n",
       "   CACO2_ENCLB422ZZZ  CACO2_ENCLB423ZZZ  SKNSH_ENCLB585ZZZ  SKNSH_ENCLB586ZZZ  \\\n",
       "0                  0                  0                  0                  0   \n",
       "1                  0                  0                  0                  0   \n",
       "2                  0                  0                  0                  0   \n",
       "3                  0                  0                  0                  0   \n",
       "4                  0                  0                  0                  0   \n",
       "\n",
       "   HMEC_ENCLB493ZZZ  GM12878_ENCLB441ZZZ  ...  fKidney_ENCLB005SRL  \\\n",
       "0                 0                    0  ...                    0   \n",
       "1                 0                    0  ...                    0   \n",
       "2                 0                    0  ...                    0   \n",
       "3                 0                    0  ...                    0   \n",
       "4                 0                    0  ...                    0   \n",
       "\n",
       "   fKidney_ENCLB704GMQ  fKidney_ENCLB759USM  fLung_ENCLB594BSZ  \\\n",
       "0                    0                    0                  0   \n",
       "1                    0                    0                  0   \n",
       "2                    0                    0                  0   \n",
       "3                    0                    0                  0   \n",
       "4                    0                    0                  1   \n",
       "\n",
       "   fKidney_ENCLB049MNH  fUmbilical_cord_ENCLB771UER  fBone_femur_ENCLB236BWV  \\\n",
       "0                    0                            0                        0   \n",
       "1                    0                            0                        0   \n",
       "2                    0                            0                        0   \n",
       "3                    0                            0                        0   \n",
       "4                    0                            0                        0   \n",
       "\n",
       "   fLiver_ENCLB638FEH  fPlacenta_ENCLB423VBC  fPlacenta_ENCLB711ZZZ  \n",
       "0                   0                      0                      0  \n",
       "1                   0                      0                      0  \n",
       "2                   0                      0                      0  \n",
       "3                   0                      0                      0  \n",
       "4                   0                      0                      0  \n",
       "\n",
       "[5 rows x 733 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binary_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4e1bd753-b08e-4bc0-8716-dce227b91db9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dhs_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>DHS_width</th>\n",
       "      <th>summit</th>\n",
       "      <th>numsamples</th>\n",
       "      <th>total_signal</th>\n",
       "      <th>component</th>\n",
       "      <th>proportion</th>\n",
       "      <th>...</th>\n",
       "      <th>fKidney_ENCLB005SRL</th>\n",
       "      <th>fKidney_ENCLB704GMQ</th>\n",
       "      <th>fKidney_ENCLB759USM</th>\n",
       "      <th>fLung_ENCLB594BSZ</th>\n",
       "      <th>fKidney_ENCLB049MNH</th>\n",
       "      <th>fUmbilical_cord_ENCLB771UER</th>\n",
       "      <th>fBone_femur_ENCLB236BWV</th>\n",
       "      <th>fLiver_ENCLB638FEH</th>\n",
       "      <th>fPlacenta_ENCLB423VBC</th>\n",
       "      <th>fPlacenta_ENCLB711ZZZ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_16140_16200_16170</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16140</td>\n",
       "      <td>16200</td>\n",
       "      <td>60</td>\n",
       "      <td>16170</td>\n",
       "      <td>1</td>\n",
       "      <td>0.129388</td>\n",
       "      <td>1</td>\n",
       "      <td>0.855153</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_51868_52040_51970</td>\n",
       "      <td>chr1</td>\n",
       "      <td>51868</td>\n",
       "      <td>52040</td>\n",
       "      <td>172</td>\n",
       "      <td>51970</td>\n",
       "      <td>1</td>\n",
       "      <td>0.080034</td>\n",
       "      <td>7</td>\n",
       "      <td>0.973545</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_57280_57354_57350</td>\n",
       "      <td>chr1</td>\n",
       "      <td>57280</td>\n",
       "      <td>57354</td>\n",
       "      <td>74</td>\n",
       "      <td>57350</td>\n",
       "      <td>4</td>\n",
       "      <td>1.093002</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_66370_66482_66430</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66370</td>\n",
       "      <td>66482</td>\n",
       "      <td>112</td>\n",
       "      <td>66430</td>\n",
       "      <td>8</td>\n",
       "      <td>1.469725</td>\n",
       "      <td>3</td>\n",
       "      <td>0.332213</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_79100_79231_79150</td>\n",
       "      <td>chr1</td>\n",
       "      <td>79100</td>\n",
       "      <td>79231</td>\n",
       "      <td>131</td>\n",
       "      <td>79150</td>\n",
       "      <td>2</td>\n",
       "      <td>0.226098</td>\n",
       "      <td>7</td>\n",
       "      <td>0.501840</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 744 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   dhs_id   chr  start    end  DHS_width  summit  numsamples  \\\n",
       "0  chr1_16140_16200_16170  chr1  16140  16200         60   16170           1   \n",
       "1  chr1_51868_52040_51970  chr1  51868  52040        172   51970           1   \n",
       "2  chr1_57280_57354_57350  chr1  57280  57354         74   57350           4   \n",
       "3  chr1_66370_66482_66430  chr1  66370  66482        112   66430           8   \n",
       "4  chr1_79100_79231_79150  chr1  79100  79231        131   79150           2   \n",
       "\n",
       "   total_signal  component  proportion  ... fKidney_ENCLB005SRL  \\\n",
       "0      0.129388          1    0.855153  ...                   0   \n",
       "1      0.080034          7    0.973545  ...                   0   \n",
       "2      1.093002          8    1.000000  ...                   0   \n",
       "3      1.469725          3    0.332213  ...                   0   \n",
       "4      0.226098          7    0.501840  ...                   0   \n",
       "\n",
       "   fKidney_ENCLB704GMQ  fKidney_ENCLB759USM  fLung_ENCLB594BSZ  \\\n",
       "0                    0                    0                  0   \n",
       "1                    0                    0                  0   \n",
       "2                    0                    0                  0   \n",
       "3                    0                    0                  0   \n",
       "4                    0                    0                  1   \n",
       "\n",
       "   fKidney_ENCLB049MNH  fUmbilical_cord_ENCLB771UER  fBone_femur_ENCLB236BWV  \\\n",
       "0                    0                            0                        0   \n",
       "1                    0                            0                        0   \n",
       "2                    0                            0                        0   \n",
       "3                    0                            0                        0   \n",
       "4                    0                            0                        0   \n",
       "\n",
       "   fLiver_ENCLB638FEH  fPlacenta_ENCLB423VBC  fPlacenta_ENCLB711ZZZ  \n",
       "0                   0                      0                      0  \n",
       "1                   0                      0                      0  \n",
       "2                   0                      0                      0  \n",
       "3                   0                      0                      0  \n",
       "4                   0                      0                      0  \n",
       "\n",
       "[5 rows x 744 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "master_dataset = pd.concat([df, binary_matrix], axis=1, sort=False)\n",
    "\n",
    "master_dataset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ccb567ab-1005-469c-b3be-fee1a837d84f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save as feather file\n",
    "# master_dataset.to_feather('master_dataset.ftr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "be0057c9-f305-4658-add1-eac588fd5ce0",
   "metadata": {},
   "outputs": [],
   "source": [
    "master_dataset = pd.read_feather('./master_dataset.ftr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "80fb27bd-483b-4471-af76-8fd0d23a34b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dhs_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>DHS_width</th>\n",
       "      <th>summit</th>\n",
       "      <th>numsamples</th>\n",
       "      <th>total_signal</th>\n",
       "      <th>component</th>\n",
       "      <th>proportion</th>\n",
       "      <th>...</th>\n",
       "      <th>fKidney_ENCLB005SRL</th>\n",
       "      <th>fKidney_ENCLB704GMQ</th>\n",
       "      <th>fKidney_ENCLB759USM</th>\n",
       "      <th>fLung_ENCLB594BSZ</th>\n",
       "      <th>fKidney_ENCLB049MNH</th>\n",
       "      <th>fUmbilical_cord_ENCLB771UER</th>\n",
       "      <th>fBone_femur_ENCLB236BWV</th>\n",
       "      <th>fLiver_ENCLB638FEH</th>\n",
       "      <th>fPlacenta_ENCLB423VBC</th>\n",
       "      <th>fPlacenta_ENCLB711ZZZ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_16140_16200_16170</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16140</td>\n",
       "      <td>16200</td>\n",
       "      <td>60</td>\n",
       "      <td>16170</td>\n",
       "      <td>1</td>\n",
       "      <td>0.129388</td>\n",
       "      <td>1</td>\n",
       "      <td>0.855153</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_51868_52040_51970</td>\n",
       "      <td>chr1</td>\n",
       "      <td>51868</td>\n",
       "      <td>52040</td>\n",
       "      <td>172</td>\n",
       "      <td>51970</td>\n",
       "      <td>1</td>\n",
       "      <td>0.080034</td>\n",
       "      <td>7</td>\n",
       "      <td>0.973545</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_57280_57354_57350</td>\n",
       "      <td>chr1</td>\n",
       "      <td>57280</td>\n",
       "      <td>57354</td>\n",
       "      <td>74</td>\n",
       "      <td>57350</td>\n",
       "      <td>4</td>\n",
       "      <td>1.093002</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_66370_66482_66430</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66370</td>\n",
       "      <td>66482</td>\n",
       "      <td>112</td>\n",
       "      <td>66430</td>\n",
       "      <td>8</td>\n",
       "      <td>1.469725</td>\n",
       "      <td>3</td>\n",
       "      <td>0.332213</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_79100_79231_79150</td>\n",
       "      <td>chr1</td>\n",
       "      <td>79100</td>\n",
       "      <td>79231</td>\n",
       "      <td>131</td>\n",
       "      <td>79150</td>\n",
       "      <td>2</td>\n",
       "      <td>0.226098</td>\n",
       "      <td>7</td>\n",
       "      <td>0.501840</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591893</th>\n",
       "      <td>chrY_56882540_56882719_56882610</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56882540</td>\n",
       "      <td>56882719</td>\n",
       "      <td>179</td>\n",
       "      <td>56882610</td>\n",
       "      <td>1</td>\n",
       "      <td>0.038079</td>\n",
       "      <td>5</td>\n",
       "      <td>0.803229</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591894</th>\n",
       "      <td>chrY_56882864_56882980_56882930</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56882864</td>\n",
       "      <td>56882980</td>\n",
       "      <td>116</td>\n",
       "      <td>56882930</td>\n",
       "      <td>1</td>\n",
       "      <td>0.115489</td>\n",
       "      <td>5</td>\n",
       "      <td>0.742349</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591895</th>\n",
       "      <td>chrY_56883733_56883960_56883830</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56883733</td>\n",
       "      <td>56883960</td>\n",
       "      <td>227</td>\n",
       "      <td>56883830</td>\n",
       "      <td>5</td>\n",
       "      <td>2.456885</td>\n",
       "      <td>7</td>\n",
       "      <td>0.559734</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591896</th>\n",
       "      <td>chrY_56884440_56884580_56884510</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56884440</td>\n",
       "      <td>56884580</td>\n",
       "      <td>140</td>\n",
       "      <td>56884510</td>\n",
       "      <td>1</td>\n",
       "      <td>0.053759</td>\n",
       "      <td>5</td>\n",
       "      <td>0.803229</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591897</th>\n",
       "      <td>chrY_56885400_56885520_56885430</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56885400</td>\n",
       "      <td>56885520</td>\n",
       "      <td>120</td>\n",
       "      <td>56885430</td>\n",
       "      <td>1</td>\n",
       "      <td>0.062719</td>\n",
       "      <td>5</td>\n",
       "      <td>0.803229</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3591898 rows × 744 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  dhs_id   chr     start       end  DHS_width  \\\n",
       "0                 chr1_16140_16200_16170  chr1     16140     16200         60   \n",
       "1                 chr1_51868_52040_51970  chr1     51868     52040        172   \n",
       "2                 chr1_57280_57354_57350  chr1     57280     57354         74   \n",
       "3                 chr1_66370_66482_66430  chr1     66370     66482        112   \n",
       "4                 chr1_79100_79231_79150  chr1     79100     79231        131   \n",
       "...                                  ...   ...       ...       ...        ...   \n",
       "3591893  chrY_56882540_56882719_56882610  chrY  56882540  56882719        179   \n",
       "3591894  chrY_56882864_56882980_56882930  chrY  56882864  56882980        116   \n",
       "3591895  chrY_56883733_56883960_56883830  chrY  56883733  56883960        227   \n",
       "3591896  chrY_56884440_56884580_56884510  chrY  56884440  56884580        140   \n",
       "3591897  chrY_56885400_56885520_56885430  chrY  56885400  56885520        120   \n",
       "\n",
       "           summit  numsamples  total_signal  component  proportion  ...  \\\n",
       "0           16170           1      0.129388          1    0.855153  ...   \n",
       "1           51970           1      0.080034          7    0.973545  ...   \n",
       "2           57350           4      1.093002          8    1.000000  ...   \n",
       "3           66430           8      1.469725          3    0.332213  ...   \n",
       "4           79150           2      0.226098          7    0.501840  ...   \n",
       "...           ...         ...           ...        ...         ...  ...   \n",
       "3591893  56882610           1      0.038079          5    0.803229  ...   \n",
       "3591894  56882930           1      0.115489          5    0.742349  ...   \n",
       "3591895  56883830           5      2.456885          7    0.559734  ...   \n",
       "3591896  56884510           1      0.053759          5    0.803229  ...   \n",
       "3591897  56885430           1      0.062719          5    0.803229  ...   \n",
       "\n",
       "        fKidney_ENCLB005SRL  fKidney_ENCLB704GMQ  fKidney_ENCLB759USM  \\\n",
       "0                         0                    0                    0   \n",
       "1                         0                    0                    0   \n",
       "2                         0                    0                    0   \n",
       "3                         0                    0                    0   \n",
       "4                         0                    0                    0   \n",
       "...                     ...                  ...                  ...   \n",
       "3591893                   0                    0                    0   \n",
       "3591894                   0                    0                    0   \n",
       "3591895                   0                    0                    0   \n",
       "3591896                   0                    0                    0   \n",
       "3591897                   0                    0                    0   \n",
       "\n",
       "         fLung_ENCLB594BSZ  fKidney_ENCLB049MNH  fUmbilical_cord_ENCLB771UER  \\\n",
       "0                        0                    0                            0   \n",
       "1                        0                    0                            0   \n",
       "2                        0                    0                            0   \n",
       "3                        0                    0                            0   \n",
       "4                        1                    0                            0   \n",
       "...                    ...                  ...                          ...   \n",
       "3591893                  0                    0                            0   \n",
       "3591894                  0                    0                            0   \n",
       "3591895                  0                    0                            0   \n",
       "3591896                  0                    0                            0   \n",
       "3591897                  0                    0                            0   \n",
       "\n",
       "         fBone_femur_ENCLB236BWV  fLiver_ENCLB638FEH  fPlacenta_ENCLB423VBC  \\\n",
       "0                              0                   0                      0   \n",
       "1                              0                   0                      0   \n",
       "2                              0                   0                      0   \n",
       "3                              0                   0                      0   \n",
       "4                              0                   0                      0   \n",
       "...                          ...                 ...                    ...   \n",
       "3591893                        0                   0                      0   \n",
       "3591894                        0                   0                      0   \n",
       "3591895                        0                   0                      0   \n",
       "3591896                        0                   0                      0   \n",
       "3591897                        0                   0                      0   \n",
       "\n",
       "         fPlacenta_ENCLB711ZZZ  \n",
       "0                            0  \n",
       "1                            0  \n",
       "2                            0  \n",
       "3                            0  \n",
       "4                            0  \n",
       "...                        ...  \n",
       "3591893                      0  \n",
       "3591894                      0  \n",
       "3591895                      0  \n",
       "3591896                      0  \n",
       "3591897                      0  \n",
       "\n",
       "[3591898 rows x 744 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "master_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "34e5fe27-dbf0-431e-99aa-274ceb587bf6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['dhs_id', 'chr', 'start', 'end', 'DHS_width', 'summit', 'numsamples',\n",
       "       'total_signal', 'component', 'proportion', 'sequence',\n",
       "       'GM06990_ENCLB435ZZZ', 'HepG2_ENCLB480ZZZ', 'hTH1_ENCLB591ZZZ',\n",
       "       'Hela_ENCLB479ZZZ', 'CACO2_ENCLB422ZZZ', 'CACO2_ENCLB423ZZZ',\n",
       "       'SKNSH_ENCLB585ZZZ', 'SKNSH_ENCLB586ZZZ', 'HMEC_ENCLB493ZZZ'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "master_dataset.columns[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5438c4ef-df54-447c-83b9-bb9d2375ee37",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "torch_new",
   "language": "python",
   "name": "torch_new"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
